gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java

```diff
@@ -172,12 +178,15 @@ public String toString() {
   @Getter
   private final Integer numThreads;
   private final Integer pollingInterval;
+  private final Integer retentionInterval;
   @Getter
   private final JobStatusRetriever jobStatusRetriever;
   private final Config config;
   private final Optional<EventSubmitter> eventSubmitter;
   private final int defaultQuota;
   private final Map<String, Integer> perUserQuota;
+  private final long failedDagRetentionTime;
+  private final Map<String, Dag<JobExecutionPlan>> failedDags = new ConcurrentHashMap<>();
```

Comment:
Why did we start creating this map here instead of inside DagManagerThread?

Reply:
The retention thread needs to delete from the same map as well, otherwise the DagManagerThread will think there is a failed dag but not find it in the failed dag state store. The failedDags map is basically a local version of the failed dag state store to avoid querying it often.